README.md
# Redefined Header Layout

This React Component allows you handle your components within a layout with a toolbar and a content.

This component makes use of [Material UI Components](https://github.com/mui-org/material-ui).

## Installation

To install this Component, run `yarn add @reactjspain/headerLayout` or `npm install @reactjspain/headerLayout`.

## Usage

To use the component, In your react Application just do

```javascript
import React from 'react'
import HeaderLayout from '@reactjspain/headerLayout'

const MyComponent = props => {

  return (
    <HeaderLayout title="Hello World!">
      {'Hello World!'}
    </HeaderLayout>
  )

}

export default MyComponent
```

You can also provide additional configuration like:

```javascript
    <HeaderLayout
      title="Hello World!"
      loading
      operations={[
        {id:'menu', icon:MenuIcon, onClick:() => window.alert('Hello World!')},
        {id:'account', image:'http://reactjspain.com/favicon.ico', right: true, small:true, to:'/account'},
        {id:'check', description:'Save', icon:Check, right:true}
      ]}
    >
      {'Hello World!'}
    </HeaderLayout>
```

CHANGELOG.md
## v0.1.0
> 2018-02-23

- **Breaking:** Added first and useless version yet ([145b46c])

[145b46c]: https://github.com/reactjspain/headerLayout/commit/145b46c

src/index.test.js
import React from 'react'
import ReactDOM from 'react-dom'
import checkPropTypes from 'check-prop-types'
import HeaderLayout from '.'
import Toolbar from 'material-ui/Divider'
import Enzyme from 'enzyme'
import { shallow, mount, render } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

Enzyme.configure({ adapter: new Adapter() })

it('renders without crashing', () => {
  const div = document.createElement('div')
  ReactDOM.render(<HeaderLayout title={'Hello World'} />, div)
})

it('when required props are not present, it shows an error', () => {
  const wrapper = shallow(<HeaderLayout classes={{}} />)
  expect(wrapper.prop('classes')).toBeDefined()
})

it('should render with Toolbar component', () => {
  const wrapper = shallow(<HeaderLayout />);
  expect(render(wrapper).find(<Toolbar />)).toBeTruthy()
})

src/index.js
import React from 'react'
import PropTypes from 'prop-types'
import CustomToolbar from './toolbar'
import { withStyles } from 'material-ui/styles'

const secondaryToolbarProperties = {
  height: 42,
  color: 'primary',
  tone: 'light',
  fontColor: 'primary',
  fontTone: 'dark'
}

const styles = theme => ({
  root: {
    display: 'flex',
    position: 'relative',
    flexDirection: 'column',
    width: '100%',
    alignItems: 'stretch',
    margin: 0,
    padding: 0
  },
  toolbar: {
    overflow: 'hidden',
    width: '100%',
    flex: '0 0 auto',
    paddingRight: 0,
    boxShadow: '0px 2px 4px -1px rgba(0, 0, 0, 0.2), 0px 4px 5px 0px rgba(0, 0, 0, 0.14), 0px 1px 10px 0px rgba(0, 0, 0, 0.12)'
  },
  content: {
    flex: '1 1 auto',
    overflowY: 'auto',
    width: '100%',
    height: 'calc(100vh - 56px)',
    marginTop: 56,
    [`${theme.breakpoints.up('xs')} and (orientation: landscape)`]: {
      marginTop: 48,
      height: 'calc(100vh - 48px)'
    },
    [theme.breakpoints.up('sm')]: {
      marginTop: 64,
      height: 'calc(100vh - 64px)'
    },
    transition: theme.transitions.create('margin-top', {
      easing: theme.transitions.easing.easeInOut,
      duration: theme.transitions.duration.shorter
    })
  }
})

const RJSHeaderLayout = props => {
  const {
    hidden,
    relative,
    relativeHeight,
    overflow,
    contentToolbar,
    secondaryToolbar,
    secondaryToolbarHeight,
    operations,
    children,
    classes,
    theme,
    ...rest 
  } = props

  let toolbarComputedStyle = {
    display: 'block',
    position:'fixed',
    zIndex: theme.zIndex.appBar+1
  }
  const secondaryToolbarProps = {
    ...secondaryToolbarProperties,
    height: secondaryToolbarHeight || secondaryToolbarProperties.height
  }
  let contentComputedStyle = {overflow : overflow || 'auto'}
  if (relative) {
    toolbarComputedStyle = {
      ...toolbarComputedStyle,
      position:'relative',
      zIndex: 0,
      boxShadow: '0px 0px'
    }
    contentComputedStyle = {
      ...contentComputedStyle,
      marginTop: 0,
      maxHeight: relativeHeight
    }
  }
  if (secondaryToolbar && !hidden) {
    contentComputedStyle = {
      height: `calc(100vh - ${secondaryToolbarProps.height}px)`,
      marginTop: secondaryToolbarProps.height,
      ...contentComputedStyle
    }
  }
  if (hidden) {
    toolbarComputedStyle = {
      display: 'none'
    }
    contentComputedStyle = {
      marginTop: 0,
      ...contentComputedStyle,
      height: '100vh'
    }
  }
  if (!children) {
    contentComputedStyle = {
      ...contentComputedStyle,
      height: 0
    }
  }

  return (
    <div className={classes.root}>

      <div className={classes.toolbar} style={toolbarComputedStyle}>
        <CustomToolbar
          secondary={secondaryToolbar}
          secondaryProps={secondaryToolbarProps}
          customContent={contentToolbar}
          operations={operations || []}
          relative={relative}
          {...rest}
        />
      </div>
      
      <div className={classes.content} style={contentComputedStyle}>
        {children}
      </div>

    </div>
  )
}

RJSHeaderLayout.propTypes = {
  classes: PropTypes.object.isRequired,
  children: PropTypes.node,
  hidden: PropTypes.bool,
  relative: PropTypes.bool,
  relativeHeight: PropTypes.number,
  operations: PropTypes.array,
  title: PropTypes.string,
  overflow: PropTypes.string,
  updateSearchQuery: PropTypes.func,
  contentToolbar: PropTypes.node,
  loading: PropTypes.bool
}

RJSHeaderLayout.defaultProps = {
  relative: false
}

export default withStyles(styles, {withTheme: true})(RJSHeaderLayout)


src/operation.js
import React from 'react'
import { Link } from 'react-router-dom'
import PropTypes from 'prop-types'
import IconButton from 'material-ui/IconButton'
import Button from 'material-ui/Button'
import Tooltip from 'material-ui/Tooltip'

const transformIdToTitle = id => id.replace(/([A-Z])/g, ' $1')

const Action = props => {

  const { id, image, to, icon, color, smallMode, small, disabled, onClick } = props

  if (image) {
    const imageStyle = {
      borderRadius: '50%',
      width: 32,
      height: 32,
      marginTop: 4,
      marginRight: 2,
      marginLeft: 2
    }
    return <img alt={id} src={image} style={imageStyle} />
  }

  const Icon = icon

  const iconButtonStyle = {
    color,
    cursor: onClick || to ? 'pointer' : 'default',
    opacity: disabled ? 0.5 : 1
  }

  const buttonStyle = {
    color,
    marginRight:2,
    marginLeft:2,
    padding:8,
    opacity: disabled ? 0.5 : 1
  }

  const iconStyle = {
    width:20,
    height:20,
    marginRight:4
  }

  const isSmall = small || smallMode
  const buttonProps = {
    disabled,
    onClick,
    style: isSmall ? iconButtonStyle : buttonStyle
  }

  if (isSmall) {
    return (
      <IconButton {...buttonProps}>
        <Icon aria-label={id} />
      </IconButton>
    )
  }

  return (
    <Button {...buttonProps}>
      <Icon style={iconStyle} aria-label={id} />
      {transformIdToTitle(id)}
    </Button>
  )
}

const Operation = props => {
  const { hidden, to, disabled, description, descriptionWhenDisabled, small, smallMode } = props
  const showDescription = 
    description && ((disabled && descriptionWhenDisabled) || (small || smallMode))

  if (hidden) {
    return <div hidden></div>
  }

  if (to) {
    return (
      <Link to={to}>
        <Action {...props} />
      </Link> 
    )
  }

  if (showDescription) {
    return (
      <Tooltip
        title={description}
        placement="left"
        disableTriggerTouch
        disableTriggerFocus
        enterDelay={1000}
        leaveDelay={0}
      >
        <div>
          <Action {...props} />
        </div>
      </Tooltip>
    )
  }

  return <Action {...props} />

}

Operation.propTypes = {
  id: PropTypes.string.isRequired,
  icon: PropTypes.func,
  image: PropTypes.string,
  color: PropTypes.string,
  hidden: PropTypes.bool,
  disabled: PropTypes.bool,
  disabledDescription: PropTypes.string,
  to: PropTypes.string,
  description: PropTypes.string,
  onClick: PropTypes.func
}

Operation.defaultProps = {
  hidden: false,
  disabled: false,
  color: 'contrast'
}

export default Operation

webpack.config.js
const webpack = require('webpack')
const path = require('path')

const BUILD_DIR = path.resolve(__dirname, 'lib')
const APP_DIR = path.resolve(__dirname, 'src')

let WebpackConfig = {

  entry: APP_DIR + '/index.js',

  output: {
    path: BUILD_DIR,
    filename: 'index.js',
    libraryTarget: 'umd',
    library: 'HeaderLayout'
  },

  module: {
    rules: [
      {
        test: /\.(js|jsx)$/,
        exclude: /node_modules/,
        include : APP_DIR,
        use: {
          loader: 'babel-loader'
          //config in .babelrc
        }
      },
      {
        test: /\.js$/,
        exclude: /node_modules/,
        include : APP_DIR,
        use: ['eslint-loader']
      }
    ]
  }

}

if ( process.env.NODE_ENV === 'production' ) {

  WebpackConfig.externals = {
    'react': 'react',
    'react-dom': 'react-dom'
  }

  WebpackConfig.plugins = [
    new webpack.optimize.AggressiveMergingPlugin(),
    new webpack.optimize.UglifyJsPlugin({
      beautify: false,
      mangle: {
        screw_ie8: true
      },
      compress: {
        warnings: false,
        screw_ie8: true
      },
      comments: false
    })
  ]
}

module.exports = WebpackConfig

src/toolbar.js
import React, { Component } from 'react'
import PropTypes from 'prop-types'
import Operation from './operation'
import { withStyles } from 'material-ui/styles'
import AppBar from 'material-ui/AppBar'
import Input from 'material-ui/Input'
import Toolbar from 'material-ui/Toolbar'
import Typography from 'material-ui/Typography'
import Search from 'material-ui-icons/Search'
import FindInPage from 'material-ui-icons/FindInPage'
import Close from 'material-ui-icons/Close'
import Reply from 'material-ui-icons/Reply'
import CircularProgress from 'material-ui/Progress/CircularProgress'
import { transformColor } from './utils/helpers'

const sizeDifferenceLimit = 'sm'

const styles = theme => ({
  /* Bar */
  appBar: {
    width: '-webkit-fill-available'
  },
  toolbar: {
    justifyContent: 'center'
  },

  /* Operations and Content */
  leftOperations: {
  },
  content: {
    display: 'flex',
    overflow: 'hidden',
    flex: 1
  },
  rightOperations: {
  },

  /* Content */
  mainInfo: {
    display: 'flex',
    alignItems: 'center',
    overflow: 'hidden',
    justifyContent: 'flex-start',
    marginRight: theme.spacing.unit
  },
  search: {
    height: '100%',
    [theme.breakpoints.up(sizeDifferenceLimit)]: {
      flex: 1,
      display: 'flex',
      justifyContent: 'flex-end',
      marginLeft: theme.spacing.unit*2,
      marginRight: theme.spacing.unit*2
    }
  },
  operations: {
    display: 'flex',
    alignItems: 'center'
  },

  /* Title */
  mainInfoText: {
    display: 'flex',
    flexDirection: 'column',
    overflow: 'hidden',
    flex: 1
  },
  mainInfoTitle: {
    overflow: 'hidden',
    textOverflow: 'ellipsis',
    whiteSpace: 'nowrap'
  },
  mainInfoDescription: {
    overflow: 'hidden',
    textOverflow: 'ellipsis',
    whiteSpace: 'nowrap',
    fontSize: 12,
    lineHeight: '1.2em'
  },
  progress: {
    margin: `0 ${theme.spacing.unit*2}px`
  },

  /* Search */
  searchBar: {
    display: 'none',
    justifyContent: 'center',
    alignItems: 'center',
    position: 'absolute',
    width: 0,
    top: theme.spacing.unit*1.5,
    left: theme.spacing.unit*2,
    zIndex: theme.zIndex.appBar,
    maxWidth: `calc(100% - ${theme.spacing.unit*10}px)`,
    transition: theme.transitions.create('all', {
      duration: theme.transitions.duration.standard,
      easing: theme.transitions.easing.sharp
    })
  },
  searchBarSearchIcon: {
    position: 'absolute',
    left: 0,
    top: '50%',
    transform: 'translate(0, -50%)',
    zIndex: 5,
    paddingLeft: theme.spacing.unit,
    paddingRight: theme.spacing.unit,
    width: theme.spacing.unit*4.5,
    height: theme.spacing.unit*4.5
  },
  searchBarInput: {
    flex: 1,
    outline: 'none',
    border: 'none',
    fontSize: 14,
    paddingLeft: theme.spacing.unit*5,
    paddingRight: theme.spacing.unit*4
  },
  searchBarCloseIcon: {
    position: 'absolute',
    right: 0,
    top: '50%',
    transform: 'translate(0, -50%)',
    paddingRight: theme.spacing.unit,
    width: theme.spacing.unit*3.5,
    height: theme.spacing.unit*3.5,
    cursor: 'pointer'
  },
  searchOperation: {
    [theme.breakpoints.up(sizeDifferenceLimit)]: {
      display: 'none'
    }
  }
})

let Operations = props => {
  const { operations, color, smallMode, classes } = props

  return (
    <div className={classes.operations}>
      {operations.map(operation => 
        <Operation key={operation.id} smallMode={smallMode} color={color} {...operation} />
      )}
    </div>
  )
}

Operations.propTypes = {
  classes: PropTypes.object.isRequired,
  operations: PropTypes.array.isRequired
}

Operations = withStyles(styles)(Operations)

class CustomToolbar extends Component {
  state = {
    searchQuery: '',
    showSearchInput: false,
    searchInputFocused: false,
    smallMode: false
  }

  componentDidMount = () => {
    window.addEventListener('resize', this.onResize)
    this.updateSmallMode(this.props.theme)
  }

  componentWillUnmount = () => {
    window.removeEventListener('resize', this.onResize)
  }

  onResize = () => {
    this.updateSmallMode(this.props.theme)
  }

  updateSmallMode = theme => {
    const width = window.innerWidth
    if (width >= theme.breakpoints.width(sizeDifferenceLimit)) {
      this.setState({smallMode: false, showSearchInput: true})
    } else {
      this.setState({smallMode: true, showSearchInput: false})
    }
  }

  updateSearchQuery = searchQuery => {
    this.setState({searchQuery})
    this.props.updateSearchQuery(searchQuery)
  }

  render = () => {
    const {
      title,
      relative,
      backgroundColor,
      description,
      operations,
      updateSearchQuery,
      customContent,
      secondary,
      secondaryProps,
      loading,
      classes,
      theme 
    } = this.props
    const { showSearchInput, searchQuery, searchInputFocused, smallMode } = this.state

    const background = backgroundColor ? backgroundColor : theme.palette.primary.main
    const color = '#FFFFFF' //TODO GETTING FROM background
    
    let appBarComputedStyle = {
      
    }
    let toolbarComputedStyle = {}
    let contentComputedStyle = {color}
    let searchBarComputedStyle = {    
      background: transformColor(background, searchInputFocused ? 24 : 16)
    }
    const descriptionComputedStyle = {
      ...contentComputedStyle,
      color: transformColor(color, -24)
    }

    /* COLORS */
    if (backgroundColor) {
      toolbarComputedStyle = {
        ...toolbarComputedStyle,
        background
      }
    }

    /* WHEN SEARCH APPEARS */
    if (showSearchInput) {
      if (smallMode) {
        searchBarComputedStyle = {
          ...searchBarComputedStyle,
          width: '100%',
          display: 'flex'
        }
      } else {
        searchBarComputedStyle = {
          ...searchBarComputedStyle,
          position: 'relative',
          display: 'flex',
          top: 0,
          left: 0,
          width: searchInputFocused || relative ? '100%' : '90%',
          maxWidth: searchInputFocused || relative ? 300 : 240
        }
        if (relative) {
          searchBarComputedStyle = {
            ...searchBarComputedStyle,
            transition: 'none'
          }
        }
      }
    }

    /* OPERATIONS */
    const leftOperations = operations.filter(operation => !operation.right)
    let rightOperations = operations.filter(operation => operation.right)

    /* IS SECONDARY TOOLBAR */
    if (secondary && secondaryProps) {
      appBarComputedStyle = {
        ...appBarComputedStyle,
        height: secondaryProps.height,
        background: theme.palette[secondaryProps.color][secondaryProps.tone]
      }
      toolbarComputedStyle = {
        ...toolbarComputedStyle,
        minHeight: secondaryProps.height,
        paddingLeft: theme.spacing.unit * (leftOperations.filter(op => !op.hidden).length ? 1 : 3),
        paddingRight: theme.spacing.unit * (rightOperations.filter(op => !op.hidden).length ? 1 : 3)
      }
      contentComputedStyle = {
        ...contentComputedStyle,
        marginBottom: 0,
        color: theme.palette[secondaryProps.fontColor][secondaryProps.fontTone]
      }
      searchBarComputedStyle = {
        ...searchBarComputedStyle,
        top: smallMode ? theme.spacing.unit/2 : 0,
        left: smallMode ? theme.spacing.unit*2 : 0,
        background: theme.palette[secondaryProps.color][secondaryProps.tone]  
      }
    }

    /* SEARCH ICON IN SMALL MODE */
    if (updateSearchQuery && smallMode) {
      if (showSearchInput) {
        rightOperations = [
          {
            id:'undo',
            description: 'Leave search',
            icon:Reply,
            color:contentComputedStyle.color,
            right:true,
            onClick:() => {
              this.setState({showSearchInput: false, searchQuery: ''})
              updateSearchQuery('')
            }
          }
        ]
      } else {
        rightOperations = [
          {
            id:'search-small',
            description: 'Search',
            icon:secondary ? FindInPage : Search,
            color:contentComputedStyle.color,
            right:true,
            onClick:() => {
              this.setState({showSearchInput: true})
              document.getElementById('searchInput').focus()
            }
          },
          ...rightOperations
        ]
      }
    }
    
    return (
      <AppBar position="static" className={classes.appBar} style={appBarComputedStyle}>
        <Toolbar disableGutters={smallMode} className={classes.appBar} style={toolbarComputedStyle}>

          <div className={classes.leftOperations}>
            <Operations
              operations={leftOperations}
              smallMode
              color={contentComputedStyle.color}
            />
          </div>

          <div className={classes.content}>
            {customContent !== undefined && customContent}
            {customContent === undefined &&
              <React.Fragment>
                {title &&
                  <div className={classes.mainInfo}>
                    <div className={classes.mainInfoText}>
                      <Typography
                        className={classes.mainInfoTitle}
                        style={contentComputedStyle}
                        type={secondary ? 'subheading' : 'title'}
                      >
                        {title}
                      </Typography>
                      <Typography
                        className={classes.mainInfoDescription}
                        style={descriptionComputedStyle}
                      >
                        {description}
                      </Typography>
                    </div>
                    {loading &&
                      <div className={classes.progress}>
                        <CircularProgress size={20} thickness={7} color="accent" />
                      </div>
                    }
                  </div>
                }

                {updateSearchQuery &&
                  <div className={classes.search}>
                    <div className={classes.searchBar} style={searchBarComputedStyle}>
                      <Search
                        className={classes.searchBarSearchIcon}
                        style={contentComputedStyle}
                      />
                      <Input
                        id="searchInput"
                        classes={{
                          root: classes.searchBarInput
                        }}
                        style={contentComputedStyle}
                        onFocus={() => this.setState({searchInputFocused: true})}
                        onBlur={() => this.setState({searchInputFocused: false})}
                        placeholder="Search"
                        disableUnderline={secondary ? false : true}
                        value={searchQuery}
                        onChange={event => this.updateSearchQuery(event.target.value)}
                      />
                      {searchQuery && 
                        <Close
                          size={20}
                          className={classes.searchBarCloseIcon}
                          style={contentComputedStyle}
                          onClick={ () => this.updateSearchQuery('')}
                        />
                      }
                    </div>
                  </div>
                }
              </React.Fragment>
            }
          </div>

          <div className={classes.rightOperations}>
            <Operations
              operations={rightOperations}
              smallMode={smallMode || relative}
              color={contentComputedStyle.color} 
            />
          </div>

        </Toolbar>
      </AppBar>
    )
  }
}

CustomToolbar.propTypes = {
  theme: PropTypes.object.isRequired,
  classes: PropTypes.object.isRequired,
  operations: PropTypes.array.isRequired,
  title: PropTypes.string,
  updateSearchQuery: PropTypes.func,
  loading: PropTypes.bool.isRequired,
  secondary: PropTypes.bool
}

CustomToolbar.defaultProps = {
  loading: false,
  secondary: false
}

export default withStyles(styles, {withTheme: true})(CustomToolbar)
